ocs/agent/aggregator.py

```diff
@@ -0,0 +1,471 @@
+        self.last_refresh = time.time() # Determines if
+
+
+    def refresh(self):
+        """Refresh provider"""
+        self.last_refresh = time.time()
+
+    def stale(self):
+        """Returns true if provider is stale and should be removed"""
+        return (time.time() - self.last_refresh) > self.fresh_time
+
+    def new_frame_time(self):
+        """Returns true if its time for a new frame to be written"""
+        if self.frame_start_time is None:
+            return False
+
+        return (time.time() - self.frame_start_time) > self.frame_length
+
+    def empty(self):
+        """Returns true if all blocks are empty"""
+        for _, b in self.blocks.items():
+            if not b.empty():
+                return False
+
+        return True
+
+    def write(self, data):
+        """
+        Saves a list of data points into blocks.
+        A block will be created for any new block_name.
+        """
+        self.refresh()
+
+        if self.frame_start_time is None:
+            # Get min frame time out of all blocks
+            self.frame_start_time = time.time()
+            for _,b in data.items():
+                if b['timestamps']:
+                    self.frame_start_time = min(self.frame_start_time, b['timestamps'][0])
+
+        for key,block in data.items():
+            try:
+                b = self.blocks[key]
+            except KeyError:
+                self.blocks[key] = ocs_feed.Block(
+                    key, block['data'].keys(),
+                    prefix=block['prefix']
+                )
+                b = self.blocks[key]
+
+            b.extend(block)
+
+    def clear(self):
+        """Clears all blocks and resets the frame_start_time"""
+        for _,b in self.blocks.items():
+            b.clear()
+
+        self.frame_start_time = None
+
+    def to_frame(self, hksess=None, clear=False):
+        """
+        Returns a G3Frame based on the provider's blocks.
+
+        Args:
+            hksess (optional):
+                If provided, the frame will be based off of hksession's data frame.
+                If the data will be put into a clean frame.
+            clear (bool):
+                Clears provider data if True.
+        """
+
+        if hksess is not None:
+            frame = hksess.data_frame(prov_id=self.prov_id)
+        else:
+            frame = core.G3Frame(core.G3FrameType.Housekeeping)
+
+        frame['address'] = self.address
+        frame['provider_session_id'] = self.sessid
+
+        for block_name, block in self.blocks.items():
+            if not block.timestamps:
+                continue
+
+            hk = so3g.IrregBlockDouble()
+            hk.prefix = block.prefix
+            hk.t = block.timestamps
+            for key, ts in block.data.items():
+                try:
+                    hk.data[key] = ts
+                except TypeError:
+                    try:
+                        hk.data[key] = [float(x) for x in ts]
```

Comment:
Yeah, point taken. I think it would be clearer to the user if this error was somehow raised by the Agent passing the value. Perhaps additional type checking can be performed in the Feeds code.

Anyway, I removed the cast attempt.